### PR TITLE
WIFI-1542 TIP implementation of new LocationChangedApImpactingEvent

### PR DESCRIPTION
**Pull request by lynnshi-design**

### CODE REVIEW

File: location-models/src/main/java/com/telecominfraproject/wlan/location/models/events/LocationChangedApImpactingEvent.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+	public LocationChangedApImpactingEvent(Location location){
+        super(location);
+    }
+    
+    /**
+     * Constructor used by JSON
+     */
+    public LocationChangedApImpactingEvent() {
+        super(null);
```
Comment (mikehansen1970) [inline location-models/src/main/java/com/telecominfraproject/wlan/location/models/events/LocationChangedApImpactingEvent.java]: This is going to call the super constructor, below, and the dereferences on location will make an NPE. Please  fix.
    public LocationChangedEvent(Location location){
        **super(location.getCustomerId(), location.getLastModifiedTimestamp(), location);**
    }

File: provisioning-sp/src/main/java/com/telecominfraproject/wlan/streams/provisioning/EquipmentConfigPushTrigger.java [hunk 2]
```diff
@@ -25,6 +25,7 @@
 import com.telecominfraproject.wlan.equipmentgateway.models.CEGWCloseSessionRequest;
 import com.telecominfraproject.wlan.equipmentgateway.models.CEGWConfigChangeNotification;
 import com.telecominfraproject.wlan.equipmentgateway.service.EquipmentGatewayServiceInterface;
+import com.telecominfraproject.wlan.location.models.events.LocationChangedApImpactingEvent;
 import com.telecominfraproject.wlan.location.models.events.LocationChangedEvent;
```
Comment (mikehansen1970) [inline provisioning-sp/src/main/java/com/telecominfraproject/wlan/streams/provisioning/EquipmentConfigPushTrigger.java:29]: Remove import of LocationChangedEvent

**Review by mikehansen1970 — CHANGES_REQUESTED** [review]
Please fix Null pointer dereference before submitting.

File: location-models/src/main/java/com/telecominfraproject/wlan/location/models/events/LocationChangedApImpactingEvent.java (hunk 1 above)
Comment (lynnshi-design) [reply location-models/src/main/java/com/telecominfraproject/wlan/location/models/events/LocationChangedApImpactingEvent.java]: @mikehansen1970 , thank you for comments. I just made a new change, to make LocationChangedApImpactingEvent not an instance of LocationChangedEvent any more, so CMAP will get two events for LocationChangedEvent handling.
With the new change, the new constructor will be a standard one, like others, which only used for JSON.

**Review by mikehansen1970 — APPROVED** [review]
Approved

File: location-service/src/main/java/com/telecominfraproject/wlan/location/service/LocationServiceController.java [hunk 3]
```diff
@@ -93,9 +94,17 @@ public Location update(@RequestBody Location location) {
             LOG.error("Failed to update location, request contains unsupported value: {}", location);
             throw new DsDataValidationException("Location contains unsupported value");
         }
+        
+        Location existingLocation = locationDatastore.get(location.getId());
 
         Location ret = locationDatastore.update(location);
         
+        if (existingLocation != null && existingLocation.getDetails() != null && ret != null && ret.getDetails() != null &&
+        		!existingLocation.getDetails().getCountryCode().equals(ret.getDetails().getCountryCode())) {
```
Comment (christieunitt-cu) [inline location-service/src/main/java/com/telecominfraproject/wlan/location/service/LocationServiceController.java:103]: Add NPE check for existingLocation.getDetails().getCountryCode() ?

**Review by christieunitt-cu — APPROVED** [review]

File: location-service/src/main/java/com/telecominfraproject/wlan/location/service/LocationServiceController.java [hunk 4]
```diff
@@ -93,9 +94,17 @@ public Location update(@RequestBody Location location) {
...
+        	LocationChangedApImpactingEvent apImpactingEvent = new LocationChangedApImpactingEvent(ret);
+            publishEvent(apImpactingEvent);
```
Comment (dmitry-toptygin-connectus-ai) [inline location-service/src/main/java/com/telecominfraproject/wlan/location/service/LocationServiceController.java:105]: when more the one event is to be published - please use the publishEvents or publishEventsBulk methods of the CloudEventDispatcherInterface

**Review by dmitry-toptygin-connectus-ai — CHANGES_REQUESTED** [review]
please address all posted comments

**Review by dmitry-toptygin-connectus-ai — APPROVED** [review]

**Pull request merged**